### ISSUE TITLE
Unable to generate recursive schema

### ISSUE THREAD

**Issue by nathan5280** [lidatong/dataclasses-json#112]
I wanted to build a recursive dataclass where a Node contains a sub Node.  It took a bit to get a workaround for the forward reference of the Node class where the annotation for the child node resolved to a class object for Node and not a string 'Node'.  You can see below how I worked that out by manually updating the __annnotation__ on the NodeRecursive class and then applying the dataclass decorator.

The problem that I've now run into is that when the schema is generated it just drills down recursively until the interpreter raises a RecursionError.  It looks like we need some way to check in mm.py build_schema(), schema() or built_type() to check if type_ == cls and stop the recursion.  At that point I think we need to add some mm field information so that it knows the field type is "self".

I'm out of time for today, but will get back to this if someone doesn't beat me to it.

There is some information in the recursive nesting in the last section of this [page](https://marshmallow.readthedocs.io/en/3.0/nesting.html).

```python
import json
from dataclasses import dataclass
from dataclasses import fields as dc_fields

from dataclasses_json import dataclass_json


@dataclass_json
@dataclass
class NodeForwardRef:
    """
    Start of the problem at hand.  The node field is of type NodeForwardRef which
    isn't defined yet.  Using a string for the type works fine for annotations
    which are analyzed by tools and not used at runtime.  At some point in the future
    this will be the default behavior.  You can experiment with it by importing
    annotations from __future__
    """
    name: str
    node: "NodeForwardRef" = None


@dataclass_json
@dataclass
class Dummy:
    """Any dataclass to use for comparision of recursive and non-recursive definitions."""
    name: str


@dataclass_json
class Node:
    """
    A class that will be manually annotated and converted to a dataclass.  The node field
    in this class will be annotated to point to a Dummy class.
    """
    # Need something in the class to trigger annotations
    name: str
    # Will manually apply the annotation for this field manually along
    # with decorating it as a dataclass.
    node = None


@dataclass_json
class NodeRecursive:
    """
    A class that will be manually annotated and converted to a dataclass.  The node field
    in this class will recursively point to another NodeRecursive.
    """
    # Need something in the class to trigger annotations
    name: str
    # Will manually apply the annotation for this field manually along
    # with decorating it as a dataclass.
    node = None


if __name__ == "__main__":
    # Create the dataclass that uses a string literal to annotate the node field.
    # This will show that the annotation resovles to a string and not to a class object.
    print("\nNodeForwardRef dataclass:")
    for field in dc_fields(NodeForwardRef):
        print("\t", field.name, field.type, type(field.type))
    print("\nNote that the type of the node field is a string not a class.")

    # Manually update the annotation on the node field on the Node class to
    # point to a Dummy class and then apply the dataclass decorator.
    # This shows that the annotation on the node field resolve to a class object.
    Node.__annotations__.update(node=Dummy)
    Node = dataclass(Node)
    print("\nNode dataclass:")
    for field in dc_fields(Node):
        print("\t", field.name, field.type, type(field.type))

    node = Node("Manual", Dummy("inner"))
    node_x = NodeForwardRef("Decorator", Node("inner"))

    # Check to see if the dataclasses_json methods were applied to the Node class from the mixin.
    # I wasn't sure if the dataclasses_json would check to make sure the class was a dataclass or not.
    # It turns out that it doesn't, which was a good thing.
    print("\nMethod difference:")
    print("\t", set(dir(Node)) ^ set(dir(NodeForwardRef)))

    dict_ = Node.schema().dump(node)
    print("\nManually annotated node with dummy node.")
    print(json.dumps(dict_, indent=2))

    # Now to the new problem.  The annotations on the NodeRecursive class can be update
    # as they were for the Node with the Dummy class.
    # Unfortunately when the schema is generated for this class it just keeps on going.
    # It looks like we need to do something like mashmallow passing fields.Nested("self", ...)
    # Reference the last section of this page: https://marshmallow.readthedocs.io/en/3.0/nesting.html
    NodeRecursive.__annotations__.update(node=NodeRecursive)
    NodeRecursive = dataclass(NodeRecursive)
    print("\nRecursive Node dataclass:")
    for field in dc_fields(NodeRecursive):
        print("\t", field.name, field.type, type(field.type))

    node = NodeRecursive("Recursive - 1 level", NodeRecursive("inner", None))
    try:
        dict_ = NodeRecursive.schema().dump(node)
        print("\nRecursive - 1 level:")
        print(json.dumps(dict_, indent=2))
    except RecursionError:
        pass
```

**Comment by nathan5280** [comment]
I guess the output might help people follow along with the code above.   

```bash
NodeForwardRef dataclass:
	 name <class 'str'> <class 'type'>
	 node NodeForwardRef <class 'str'>

Note that the type of the node field is a string not a class.

Node dataclass:
	 name <class 'str'> <class 'type'>
	 node <class '__main__.Dummy'> <class 'type'>

Method difference:
	 set()

Manually annotated node with dummy node.
{
  "node": {
    "name": "inner"
  },
  "name": "Manual"
}

Recursive Node dataclass:
	 name <class 'str'> <class 'type'>
	 node <class '__main__.NodeRecursive'> <class 'type'>
```
The end of the stack trace.
```bash
  File "/Users/nateatkins/projects/cake/cdst/dataclasses-json/dataclasses_json/mm.py", line 262, in schema
    t = build_type(type_, options, mixin, field, cls)
  File "/Users/nateatkins/projects/cake/cdst/dataclasses-json/dataclasses_json/mm.py", line 234, in build_type
    return inner(type_, options)
  File "/Users/nateatkins/projects/cake/cdst/dataclasses-json/dataclasses_json/mm.py", line 218, in inner
    return TYPES[origin](*args, **options)
  File "/Users/nateatkins/.local/share/virtualenvs/util-dataclass-package-Gdf7dWv4/lib/python3.7/site-packages/marshmallow/fields.py", line 149, in __init__
    if utils.is_iterable_but_not_string(validate):
  File "/Users/nateatkins/.local/share/virtualenvs/util-dataclass-package-Gdf7dWv4/lib/python3.7/site-packages/marshmallow/utils.py", line 58, in is_iterable_but_not_string
    (isinstance(obj, Iterable) and not hasattr(obj, 'strip')) or is_generator(obj)
  File "/Users/nateatkins/.local/share/virtualenvs/util-dataclass-package-Gdf7dWv4/bin/../lib/python3.7/abc.py", line 139, in __instancecheck__
    return _abc_instancecheck(cls, instance)
RecursionError: maximum recursion depth exceeded in comparison
```

**Comment by nathan5280** [comment]
I have a prototype working for this where you specify the maximum recursion depth to build the schema.   It needs so work, but solves my immediate problem.   It looks like Mashmallow has the same issue.

**Comment by nathan5280** [comment]
I added a pull request for some work I've done on this.  (https://github.com/lidatong/dataclasses-json/pull/113)

**Comment by nathan5280** [comment]
I think we can close this.  I found an example of the functionality already working in test_recursive.   I'm not sure why I couldn't figure this out before, but that is the way it goes sometimes.   I submitted a pull request for an additional test case in test_recursive to make sure it works correctly when you go against the schema() instead of just to_dict/json and from_dict/json.

```python
    def test_tree_schema_round_trip(self):
        tree_dict = Tree.schema().dump(family_tree)
        tree_obj = Tree.schema().load(tree_dict)
        assert tree_obj == family_tree
```
I also added a little blurb on this in the README so that hopefully I can spare someone else going down the rabbit hole I did.
